File: RemoveNthNodeFromEndOfList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *p1 = head, *p2 = head, *p3 = NULL;
        while (p1 != NULL) {
            p1 = p1->next;
            if (--n < 0) {
                p3 = p2;
                p2 = p2->next;
            }
        }
        if (p3 != NULL) {
            ListNode *tmp = p3->next;
            p3->next = p2->next;
            delete tmp;
        } else { 
            ListNode *tmp = head;
            head = p2->next;
            delete tmp;
        }
        
        return head;
    }
};

File: lulu/removeNthNodeFromEndOfList.py
class Solution(object):
    def remove(self, head, n):
        if head is None:
            return
        p = head
        q = head
        count = 1
        while p and count != n+1:
            p = p.next
            count += 1
        
        if p is None:
            return head.next
        else:
            while p and p.next:
                p = p.next
                q = q.next
            

            q.next = q.next.next
            return head
        
